tool/tsh/kube.go

```diff
@@ -0,0 +1,269 @@
+			log.Debugf("Re-using existing TLS cert for kubernetes cluster %q", c.kubeCluster)
+			return c.writeResponse(k, c.kubeCluster)
+		}
+		// Otherwise, cert for this k8s cluster is missing or expired. Request
+		// a new one.
+	}
+
+	log.Debugf("Requesting TLS cert for kubernetes cluster %q", c.kubeCluster)
+	err = client.RetryWithRelogin(cf.Context, tc, func() error {
+		return tc.ReissueUserCerts(cf.Context, client.ReissueParams{
+			RouteToCluster:    c.teleportCluster,
+			KubernetesCluster: c.kubeCluster,
+		})
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// ReissueUserCerts should cache the new cert on disk, re-read them.
+	k, err = tc.LocalAgent().GetKey(client.WithKubeCerts(c.teleportCluster))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return c.writeResponse(k, c.kubeCluster)
+}
+
+func (c *kubeCredentialsCommand) writeResponse(key *client.Key, kubeClusterName string) error {
+	crt, err := key.KubeTLSCertificate(kubeClusterName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	resp := &clientauthentication.ExecCredential{
+		Status: &clientauthentication.ExecCredentialStatus{
+			// Indicate  slightly earlier expiration to avoid the cert expiring
+			// mid-request.
+			ExpirationTimestamp:   &metav1.Time{Time: crt.NotAfter.Add(-1 * time.Minute)},
+			ClientCertificateData: string(key.KubeTLSCerts[kubeClusterName]),
+			ClientKeyData:         string(key.Priv),
+		},
+	}
+	data, err := runtime.Encode(kubeCodecs.LegacyCodec(kubeGroupVersion), resp)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	fmt.Println(string(data))
+	return nil
+}
+
+type kubeClustersCommand struct {
+	*kingpin.CmdClause
+}
+
+func newKubeClustersCommand(parent *kingpin.CmdClause) *kubeClustersCommand {
+	c := &kubeClustersCommand{
+		CmdClause: parent.Command("clusters", "Get credentials for kubectl access"),
+	}
+	return c
+}
+
+func (c *kubeClustersCommand) run(cf *CLIConf) error {
+	tc, err := makeClient(cf, true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var currentTeleportCluster string
+	var kubeClusters []string
+	err = client.RetryWithRelogin(cf.Context, tc, func() error {
+		pc, err := tc.ConnectToProxy(cf.Context)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer pc.Close()
+		ac, err := pc.ConnectToCurrentCluster(cf.Context, true)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer ac.Close()
+
+		cn, err := ac.GetClusterName()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		currentTeleportCluster = cn.GetClusterName()
+
+		kubeClusters, err = kubeutils.KubeClusterNames(cf.Context, ac)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var selectedCluster string
+	if kc, err := kubeconfig.Load(""); err != nil {
+		log.WithError(err).Warning("Failed parsing existing kubeconfig")
+	} else {
+		selectedCluster = strings.TrimPrefix(kc.CurrentContext, currentTeleportCluster+"-")
```

Comment:
Nit: This logic of converting teleport cluster name and k8s cluster name into a context name that's based on concatenating them with the dash is used in multiple places "verbatim" so it might be pretty fragile. I would consider making a couple of 1-liner helper methods somewhere in lib/kube that deal with it and using where it's needed.

Reply:
done